### PR TITLE
handle highlightFirstSuggestion state changes

### PR DESCRIPTION
**Pull request by DoloMike**
This is to resolve 

> [Bug] highlightedSectionIndex does not reset to null when highlightFirstSuggestion changes from true to false #759

Code was added inside `Autosuggest.js` -> `UNSAFE_componentWillReceiveProps()` to check if `nextProps.highlightFirstSuggestion` had changed. If `nextProps.highlightFirstSuggestion === false` then `this.resetHighlightedSuggestion()` is called. It ensures that nothing is highlighted once `highlightFirstSuggestion` changes from true to false.

A test was added for it under `./focus-first-autosuggestion`. All other tests remain unchanged and continue to pass.

### CODE REVIEW

File: src/Autosuggest.js
Comment (pimterry) [inline src/Autosuggest.js:140]: If `highlightFirstSuggestion` is always false, this will reset the highlighted suggestion any time any other prop changes, which doesn't seem quite right.

I think it should only reset if all of the below are true:

* The new prop sets highlightFirstSuggestion to false
* The old prop had highlightFirstSuggestion set to true
* The first suggestion is currently highlighted

For example, if the first suggestion is highlighted, I manually select the 3rd item, and then `highlightFirstSuggestion` becomes false, I think it's probably better for the suggestion to stay where it was.

That should still work for your case, but ensures this behaviour is limited just to that specific issue, and doesn't affect normal suggestion selection elsewhere.

**Review by pimterry — CHANGES_REQUESTED** [review]
Thanks for the great contribution! This seems logical, the tests & code look good, and sounds like a sensible improvement to me. It would be useful to be a little more precise about the behaviour, just to limit side effects, but once that's done I think this is good to merge.

File: test/focus-first-suggestion/AutosuggestApp.test.js
Comment (hugoholgersson) [inline test/focus-first-suggestion/AutosuggestApp.test.js:151]: I understand the commit hook reformatted these lines... Not your fault... Still, to make the history digging and reviewing easier, I think it's better to split reformatting out to its own commit.

**Comment by pimterry** [comment]
Hi @DoloMike any you still interested in this? I'd be happy to merge it if you've got a minute to handle the comments above, if not no worries, but I'll close this PR in a few days.

**Comment by DoloMike** [comment]
Yeah I have been a little tied up but I'll try to handle this week. Thank
you for the great feedback btw.

File: src/Autosuggest.js
Comment (DoloMike) [reply src/Autosuggest.js:140]: Thanks for the detailed feedback. I agree and have updated the code now.

File: test/focus-first-suggestion/AutosuggestApp.test.js
Comment (DoloMike) [reply test/focus-first-suggestion/AutosuggestApp.test.js:151]: I appreciate the feedback. I agree. I've went ahead and done that with my latest push for what it's worth.

**Comment by DoloMike** [comment]
@pimterry Sorry that I just got around to implementing the changes, but they're here. Please let me know if you would like to discuss further. I'll resolve for now.

File: src/Autosuggest.js
Comment (pimterry) [inline src/Autosuggest.js:147]: It looks like you've added the discussed change here, but also left the previous code too. Is that intentional? I think the `else if` case above shouldn't be here now, unless I'm missing something.

File: test/focus-first-suggestion/AutosuggestApp.js
Comment (pimterry) [inline test/focus-first-suggestion/AutosuggestApp.js:20]: This doesn't make much sense I'm afraid, I think there's a misunderstanding somewhere.

This code defines a function, which you manually call in the tests, but it doesn't have any side effects, so that doesn't do anything. You then pass this function as the `highlightFirstSuggestion` prop below, which is interpreted as truthy, because it's a function, but it's actually never called there. Maybe `sinon.spy` is creating some confusion: all that does is record every call to this function, so you can check in tests whether the function was called.

If you replace this with just `getSuggestionValue = () => {}` it still does the same thing (and the tests still pass), and similarly if you keep this but swap the `true`/`false` values in the test that passes too.

I think you want to toggle the `highlightFirstSuggestion` prop, so that you actually pass `true` or `false` to `<Autosuggest>` below, and control it from the tests, right? You can do that by:

* Exporting a function here that runs `app.setState({ highlightFirstSuggestion: /* true / false */})`
* Using `this.state.highlightFirstSuggestion` as the `highlightFirstSuggestion` prop in `render()` below
* Updating the test code to call your exported function when you want to switch the values.

Does that make sense? Happy to answer any questions.

**Review by pimterry — CHANGES_REQUESTED** [review]

**Review by pimterry — APPROVED** [review]
I've fixed up and finished the last few changes from my comments above myself, plus a bonus fix to match issues that the new tests discovered.

Thanks for the PR @DoloMike! This is all good now imo, so I'm going to merge & release :tada:

**Pull request merged**
